File: src/lib/delivery-fees.ts
import type { CheckoutFormData } from "@/lib/order";

export interface DemoDeliveryFee {
  neighborhood: string;
  fee: number;
}

export const DEFAULT_DEMO_DELIVERY_FEE = 8;

export const DEMO_DELIVERY_FEES: DemoDeliveryFee[] = [
  { neighborhood: "Centro", fee: 6 },
  { neighborhood: "Santo Antônio", fee: 7 },
  { neighborhood: "Cidade Nova", fee: 9 },
  { neighborhood: "Parque Piauí I", fee: 8 },
  { neighborhood: "Parque Piauí II", fee: 8 },
  { neighborhood: "Parque Alvorada", fee: 9 },
  { neighborhood: "Parque São Francisco", fee: 10 },
  { neighborhood: "Parque São Francisco II", fee: 10 },
  { neighborhood: "Parque União", fee: 10 },
  { neighborhood: "Boa Vista", fee: 9 },
  { neighborhood: "Conjunto Boa Vista", fee: 10 },
  { neighborhood: "Coheb", fee: 8 },
  { neighborhood: "Formosa", fee: 10 },
  { neighborhood: "Planalto Formosa", fee: 11 },
  { neighborhood: "Flores", fee: 12 },
  { neighborhood: "Reserva das Flores", fee: 12 },
  { neighborhood: "Guarita", fee: 8 },
  { neighborhood: "Mateuzinho", fee: 10 },
  { neighborhood: "Parque Aliança", fee: 11 },
  { neighborhood: "Pedro Patrício", fee: 12 },
  { neighborhood: "Mutirão", fee: 10 },
  { neighborhood: "Centro Operário", fee: 9 },
  { neighborhood: "Loteamento Boa Vista", fee: 11 },
  { neighborhood: "Novo Joia", fee: 12 },
];

const feeByNeighborhood = new Map(
  DEMO_DELIVERY_FEES.map((item) => [normalizeNeighborhood(item.neighborhood), item.fee]),
);

export function calculateDemoDeliveryFee(checkout: CheckoutFormData) {
  if (checkout.fulfillment !== "delivery") {
    return 0;
  }

  const neighborhood = normalizeNeighborhood(checkout.neighborhood);

  if (!neighborhood) {
    return 0;
  }

  return feeByNeighborhood.get(neighborhood) ?? DEFAULT_DEMO_DELIVERY_FEE;
}

export function normalizeNeighborhood(value: string) {
  return value
    .normalize("NFD")
    .replace(/[\u0300-\u036f]/g, "")
    .trim()
    .toLowerCase()
    .replace(/\s+/g, " ");
}

File: backend/app/services/delivery_fees.py
from __future__ import annotations

from decimal import Decimal
import unicodedata

from app.models.enums import FulfillmentType

DEFAULT_DEMO_DELIVERY_FEE = Decimal("8.00")

DEMO_DELIVERY_FEES: dict[str, Decimal] = {
    "Centro": Decimal("6.00"),
    "Santo Antônio": Decimal("7.00"),
    "Cidade Nova": Decimal("9.00"),
    "Parque Piauí I": Decimal("8.00"),
    "Parque Piauí II": Decimal("8.00"),
    "Parque Alvorada": Decimal("9.00"),
    "Parque São Francisco": Decimal("10.00"),
    "Parque São Francisco II": Decimal("10.00"),
    "Parque União": Decimal("10.00"),
    "Boa Vista": Decimal("9.00"),
    "Conjunto Boa Vista": Decimal("10.00"),
    "Coheb": Decimal("8.00"),
    "Formosa": Decimal("10.00"),
    "Planalto Formosa": Decimal("11.00"),
    "Flores": Decimal("12.00"),
    "Reserva das Flores": Decimal("12.00"),
    "Guarita": Decimal("8.00"),
    "Mateuzinho": Decimal("10.00"),
    "Parque Aliança": Decimal("11.00"),
    "Pedro Patrício": Decimal("12.00"),
    "Mutirão": Decimal("10.00"),
    "Centro Operário": Decimal("9.00"),
    "Loteamento Boa Vista": Decimal("11.00"),
    "Novo Joia": Decimal("12.00"),
}

def calculate_delivery_fee(
    *,
    fulfillment_type: FulfillmentType,
    neighborhood: str | None,
) -> Decimal:
    if fulfillment_type != FulfillmentType.DELIVERY:
        return Decimal("0.00")

    normalized_neighborhood = normalize_neighborhood(neighborhood or "")

    if not normalized_neighborhood:
        return Decimal("0.00")

    return NORMALIZED_DEMO_DELIVERY_FEES.get(
        normalized_neighborhood,
        DEFAULT_DEMO_DELIVERY_FEE,
    )


def normalize_neighborhood(value: str) -> str:
    normalized = unicodedata.normalize("NFKD", value.strip().lower())
    without_accents = "".join(
        character for character in normalized if not unicodedata.combining(character)
    )
    return " ".join(without_accents.split())


NORMALIZED_DEMO_DELIVERY_FEES = {
    normalize_neighborhood(neighborhood): fee
    for neighborhood, fee in DEMO_DELIVERY_FEES.items()
}
